### DISCUSSION TITLE
`ErrorResponse` conflicts with `impl IntoResponse`

### DISCUSSION THREAD

**Discussion by scotow** [tokio-rs/axum#1062] [Q&A] [answered]
Since the merge of this [PR](https://github.com/tokio-rs/axum/pull/921), it has become more difficult to use the `impl IntoReponse` syntax (as highly presented in the doc). It was pretty useful to not write the whole response type.

Here is an example of an over simplified handler:

```rust
async fn handler() -> impl IntoResponse {
    Err("error")?;
    Ok("success")
}
```

and here is the associated error:

```
error[E0283]: type annotations needed
  --> src/main.rs:40:41
   |
40 |   async fn handler() -> impl IntoResponse {
   |  _________________________________________^
41 | |     Err("error")?;
42 | |     Ok("success")
43 | | }
   | |_^ cannot infer type for enum `Result<&str, _>`
   |
   = note: multiple `impl`s satisfying `Result<&str, _>: IntoResponse` found in the `axum_core` crate:
           - impl<T, E> IntoResponse for Result<T, E>
             where T: IntoResponse, E: IntoResponse;
           - impl<T> IntoResponse for Result<T, ErrorResponse>
             where T: IntoResponse;
```

In this example, the return type can be specified as `Result<&'static str, &'static str>`, but when working with complex response types, writing the full types can be a pain.

Am I missing something, or is not possible to use the `impl IntoResponse` for returning `Result<_, _>` anymore?

**Comment by jplatte** [comment, marked as answer]
I'm surprised this ever worked. I would recommend writing the return types of fallible handler functions as `Result<impl IntoResponse, Error>` or `Result<impl IntoResponse>` given a `Result` type alias like the one in `axum::response`.
